### PR TITLE
Mithril: Add request config options param.

### PR DESCRIPTION
**Pull request by spacejack**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/spacejack/mithril.d.ts/issues/25#issuecomment-318577656

This PR adds an options parameter to the RequestOptions config callback.

Test/lint rules have changed since the last PR so I also needed to make some changes to tsconfig, tslint and some of the tests.

Suppressing "any index" errors is now forbidden so I removed a number of extraneous stream tests for Fantasy Land spec compliance that aren't covered by the type defs.

The new linting rules no longer like this definition:

```typescript
declare function jsonp<T>(options: Mithril.JsonpOptions & { url: string }): Promise<T>;
```

So for now I've suppressed `no-unnecessary-generics` inline. I don't see how else to describe the type of Promise you expect back.

### CODE REVIEW

**Comment by isiahmeadows** [comment]
LGTM.

File: types/mithril/test/test-stream.ts [hunk 1]
```diff
@@ -215,112 +204,6 @@ import { Stream } from 'mithril/stream';
 	console.assert(doubled() === 4);
 }
 
-{
```
Comment (aozgaa) [inline types/mithril/test/test-stream.ts]: Why were these removed?

Comment (spacejack) [reply types/mithril/test/test-stream.ts]: Those tests were copied/adapted from Mithril's own test suite, which are used to test that the fantasy land method names (which are there for spec compliance) work.

We have not added type definitions for the quoted methods (in general, users use `stream.map` not `stream["fantasy-land/map"]`, etc.)

Because the DT rules now prevent us from allowing any index types, we can't keep these tests without adding those types. And those tests weren't really useful as type tests previously.

Comment (aozgaa) [reply types/mithril/test/test-stream.ts]: Thanks for the clarification!
